File: sample-chat/src/main/java/com/quickblox/sample/chat/ui/widget/AttachmentPreviewAdapterView.java
package com.quickblox.sample.chat.ui.widget;

import android.content.Context;
import android.database.DataSetObserver;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Adapter;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;

public class AttachmentPreviewAdapterView extends HorizontalScrollView {

    private static final Handler mainThreadHandler = new Handler(Looper.getMainLooper());

    private LinearLayout container;
    private Adapter adapter;
    private DataSetObserver dataSetObserver;

    public AttachmentPreviewAdapterView(Context context, AttributeSet attrs) {
        super(context, attrs);

        container = new LinearLayout(context);
        container.setOrientation(LinearLayout.HORIZONTAL);
        HorizontalScrollView.LayoutParams lp = new HorizontalScrollView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        addView(container, lp);

        dataSetObserver = new DataSetObserver() {
            @Override
            public void onChanged() {
                populateWithViewsFromAdapter();
            }

            @Override
            public void onInvalidated() {
                populateWithViewsFromAdapter();
            }
        };
    }

    public void setAdapter(Adapter newAdapter) {
        if (adapter != null) {
            adapter.unregisterDataSetObserver(dataSetObserver);
        }
        adapter = newAdapter;
        adapter.registerDataSetObserver(dataSetObserver);
        populateWithViewsFromAdapter();
    }

    private void populateWithViewsFromAdapter() {
        mainThreadHandler.post(new Runnable() {
            @Override
            public void run() {
                container.removeAllViews();
                for (int i = 0; i < adapter.getCount(); i++) {
                    View childView = adapter.getView(i, null, AttachmentPreviewAdapterView.this);
                    container.addView(childView, i);
                }
            }
        });
    }
}

File: sample-chat-kotlin/src/main/java/com/quickblox/sample/chat/kotlin/ui/views/AttachmentPreviewAdapterView.kt
package com.quickblox.sample.chat.kotlin.ui.views

import android.content.Context
import android.database.DataSetObserver
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.Adapter
import android.widget.HorizontalScrollView
import android.widget.LinearLayout

class AttachmentPreviewAdapterView(context: Context, attrs: AttributeSet) : HorizontalScrollView(context, attrs) {
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    private val container: LinearLayout = LinearLayout(context)
    private var adapter: Adapter? = null
    private val dataSetObserver: DataSetObserver

    init {
        container.orientation = LinearLayout.HORIZONTAL
        val width = ViewGroup.LayoutParams.MATCH_PARENT
        val height = ViewGroup.LayoutParams.MATCH_PARENT
        val layoutParams = LayoutParams(width, height)
        addView(container, layoutParams)
        dataSetObserver = object : DataSetObserver() {
            override fun onChanged() {
                populateWithViewsFromAdapter()
            }

            override fun onInvalidated() {
                populateWithViewsFromAdapter()
            }
        }
    }

    fun setAdapter(newAdapter: Adapter) {
        adapter?.unregisterDataSetObserver(dataSetObserver)
        adapter = newAdapter
        adapter?.registerDataSetObserver(dataSetObserver)
        populateWithViewsFromAdapter()
    }

    private fun populateWithViewsFromAdapter() {
        mainThreadHandler.post {
            container.removeAllViews()
            adapter?.let {
                for (index in 0 until it.count) {
                    val childView = it.getView(index, null, this)
                    container.addView(childView, index)
                }
            }
        }
    }
}